### PR TITLE
[doc] Run A Flux Minicluster

### LINKED ISSUE #716: Support flux-operator
**What would you like to be added**:

Support queuing of ephemeral and persistent mini-clusters from [flux-operator](https://flux-framework.org/flux-operator/).

It is possible that this works out-of-the-box, given that flux-operator uses the Job API to launch Pods.

If it works, we need to document any implications.
If it doesn't work, maybe we need to add a suspend flag and a translation controller.

**Why is this needed**:

To accommodate multiple HPC researchers in a kubernetes cluster

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [x] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by trasc**
#### What type of PR is this?

/kind documentation

#### What this PR does / why we need it:

Documents flux MiniCluster usage with Kueue. 

#### Which issue(s) this PR fixes:

Fixes #716

### CODE REVIEW

**Comment by vsoch** [comment]
Thanks @trasc ! Just to keep you in the loop, the work to remove the cert-generator pod is pretty much done - https://github.com/flux-framework/flux-operator/pull/152 - along with removing the pod it [makes the entire process ~3 seconds faster](https://github.com/flux-framework/flux-operator/issues/150#issuecomment-1524878261). I should be able to give this one more final look over, and merge tomorrow (if you'd like to test!) so this tutorial can go in without the note about the extra pod.

It's 1:30am here and I should probably be off to bed, but will check this out tomorrow - really excited to try it! Thank you for pinging me, and good night! :zzz:

**Comment by trasc** [comment]
> Thanks @trasc ! Just to keep you in the loop, the work to remove the cert-generator pod is pretty much done - [flux-framework/flux-operator#152](https://github.com/flux-framework/flux-operator/pull/152) - along with removing the pod it [makes the entire process ~3 seconds faster](https://github.com/flux-framework/flux-operator/issues/150#issuecomment-1524878261). I should be able to give this one more final look over, and merge tomorrow (if you'd like to test!) so this tutorial can go in without the note about the extra pod.
> 
> It's 1:30am here and I should probably be off to bed, but will check this out tomorrow - really excited to try it! Thank you for pinging me, and good night! zzz

Hi, you are making progress on the issues, I think @alculquicondor can agree that we are not  in a hurry with this and can hold the merge for a couple of extra days and hopefully we cam merge  it without the "Know limitations" section.

**Comment by alculquicondor** [comment]
Yes, we can hold the documentation PR for a couple of weeks. Thank you all!

File: site/content/en/docs/tasks/run_flux.md
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux.md:1]: run_flux_minicluster.md

File: site/content/en/docs/tasks/run_flux.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+title: "Run A Flux Minicluster"
+date: 2022-02-14
+weight: 6
+description: >
+  Run a Kueue scheduled Flux minicluster.
+---
+
+This page shows how to leverage Kueue's scheduling and resource management capabilities when running [Flux Operator's](https://flux-framework.org/flux-operator/) Mini Clusters.
+
+For more information on Flux Operator visit [Flux Operator](https://flux-framework.org/flux-operator/).
```
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux.md:11]: I think we can remove this line, given that there is already a link above.

But we should add:

The intended audience for this page are [batch users](https://kueue.sigs.k8s.io/docs/tasks#batch-user).

File: site/content/en/docs/tasks/run_flux.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+
+## Before you begin
+
+Please check [Run A Job](/docs/tasks/run_jobs) for details on the initial cluster setup and queue configuration.
```
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux.md:16]: Rather, point to `/docs/tasks/administer_cluster_quotas/`

File: site/content/en/docs/tasks/run_flux.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+## 1. MiniCluster definition
+
+Since a Flux MiniCluster uses  a `batch/job` to run the actual workload, no extra components are needed for it to be managed by Kueue, However the following aspects should be taken into account:
```
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux.md:20]:
```suggestion
Since a Flux MiniCluster runs as a [`batch/Job`](https://kubernetes.io/docs/concepts/workloads/controllers/job/), no extra components are needed for it to be managed by Kueue. However, the following aspects should be taken into account:
```

File: site/content/en/docs/tasks/run_flux.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+### a. Queue selection
+
+The target local queue should be specified in the `spec.jobLabels` section of the minicluster configuration.
```
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux.md:24]: link to /docs/concepts/local_queue

File: site/content/en/docs/tasks/run_flux.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+spec:
+  containers:
+    - image: <image>
+      resources:
+        requests:
+          cpu: 4
+          memory: "200Mi"
+```
+
+## 2. Configuration example
```
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux.md:45]:
```suggestion
## 2. Sample MiniCluster
```

File: site/content/en/docs/tasks/run_flux.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        requests:
+          cpu: 4
+          memory: "200Mi"
+  jobLabels:
+    kueue.x-k8s.io/queue-name: user-queue
+```
+
+## 3. Known Limitations
+
+- The certificate generation step of the MiniCluster life cycle runs outside it's main job and it's execution will not be managed by Kueue.
```
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux.md:67]: /hold
we might be able to remove this soon.

File: site/content/en/docs/tasks/run_flux.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+- In some configuration variants the MiniCluster's job might not run to completion blocking the resources of the associated queue until it's deletion.
```
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux.md:68]:
```suggestion
- In some configuration variants, the MiniCluster's job might not run to completion blocking the resources of the associated queue until it's deletion.
```

I don't think this will change anytime soon, or is it @vsoch?

File: site/content/en/docs/tasks/run_flux.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+```yaml
+apiVersion: flux-framework.org/v1alpha1
+kind: MiniCluster
+metadata:
+  generateName: flux-sample-kueue-
+spec:
+  size: 1
+  containers:
+    - image: ghcr.io/flux-framework/flux-restful-api:latest
+      command: echo hello world
```
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux.md:56]: Is this a meaningful command @vsoch?

Is there a hello HPC? calculate pi, etc?

**Review by alculquicondor — COMMENTED** [review]
@cortespao could you have a quick look?

File: site/content/en/docs/tasks/run_flux.md (hunk 2 above)
Comment (alculquicondor) [reply site/content/en/docs/tasks/run_flux.md:16]: oh, and another requirement is that the flux operator is installed :)

File: site/content/en/docs/tasks/run_flux.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+The target local queue should be specified in the `spec.jobLabels` section of the minicluster configuration.
+
+```yaml
+  jobLabels:
+    kueue.x-k8s.io/queue-name: user-queue
+```
+
+### b. Configure the resource needs
+
+The resource needs of the workload can be configured in the `spec.container[].resources` sections of the minicluster configuration.
```
Comment (vsoch) [inline site/content/en/docs/tasks/run_flux.md:33]: If the user is running something with MPI or similar and they want more control, they also might want to consider tweaking `spec.container[].cores`.  or `spec.tasks` https://flux-framework.org/flux-operator/getting_started/custom-resource-definition.html?h=tasks#tasks. In layman's terms these are telling the exact resources that should be used for the single command. As an example from our kubecon experiment run, on a size 64 cluster:

```yaml
apiVersion: flux-framework.org/v1alpha1
kind: MiniCluster

metadata:
  name: lammps
  namespace: flux-operator
spec:
  # localDeploy needs to be false
  localDeploy: false

  # Number of pods to create for MiniCluster
  size: 64
  tasks: 6016

  # don't suppress output (in case there is issue) and time flux
  logging:
    quiet: false
    timed: true
    debug: true

  # This is a list because a pod can support multiple containers
  containers:
    - image: ghcr.io/rse-ops/spack-ubuntu-libfabric-ssh:ubuntu-20.04
      workingDir: /home/flux/examples/reaxff/HNS
      command: lmp -v x 64 -v y 16 -v z 16 -in in.reaxc.hns -nocite

      # Option Flags for this flux runner wait.sh entrypoint
      fluxOptionFlags: "-ompi=openmpi@5 -c 1 -o cpu-affinity=per-task" 

      # Leave 2 cores for kubernetes, we have 96,
      # this is just for eksctl
      cores: 96

      # Resource limits to enable efa
      resources:
        limits:
          vpc.amazonaws.com/efa: 1
          memory: "340G"
          cpu: 94

        requests:
          vpc.amazonaws.com/efa: 1
          memory: "340G"
          cpu: 94

      # custom preCommand logic (run at start of script)
      preCommand: |
        source /etc/profile.d/z10_spack_environment.sh
        asFlux="sudo -u flux -E PYTHONPATH=$PYTHONPATH -E PATH=$PATH -E FI_EFA_USE_DEVICE_RDMA=1 -E RDMAV_FORK_SAFE=1"
        . /etc/profile.d/z10_spack_environment.sh 
        cd /opt/spack-environment
        . /opt/spack-environment/spack/share/spack/setup-env.sh
        spack env activate .
        cd /home/flux/examples/reaxff/HNS
```
That generated the following command for flux start:

```bash
🌀 flux start $brokerOptions flux mini submit  -N 64 -n 6016 --quiet -ompi=openmpi@5 -c 1 -o cpu-affinity=per-task --watch -vvv lmp -v x 64 -v y 16 -v z 16 -in in.reaxc.hns -nocite
```
Note that `-n` is 6016 and `-N` is 64, and in flux this maps to:

```bash
  -N, --nodes=N               Number of nodes to allocate
  -n, --ntasks=N              Number of tasks to start
```

File: site/content/en/docs/tasks/run_flux.md (hunk 8 above)
Comment (vsoch) [reply site/content/en/docs/tasks/run_flux.md:56]: Oh yeah you mean:

```bash
echo hello hpc :wave: 
```
:laughing: 

We have quite a few examples https://github.com/flux-framework/flux-operator/tree/main/examples and actually the one I'd recommend (with a pre-built container that does something marginally interesting that uses MPI) is... LAMMPS!

https://github.com/flux-framework/flux-operator/blob/main/examples/tests/lammps/minicluster.yaml

We do have a hello-world example https://github.com/flux-framework/flux-operator/blob/main/examples/tests/hello-world/minicluster.yaml but I'd stay away from it, because it runs so quickly you'd probably miss it.

File: site/content/en/docs/tasks/run_flux.md (hunk 6 above)
Comment (vsoch) [reply site/content/en/docs/tasks/run_flux.md:67]: This has been merged so safe to remove. Anyone that installs the operator after this morning will not get a certificate pod!

File: site/content/en/docs/tasks/run_flux.md (hunk 7 above)
Comment (vsoch) [reply site/content/en/docs/tasks/run_flux.md:68]: I'm definitely up to try out different strategies - what we really need is to have the logic afforded by the new JobSet. My only idea was to go off of what you mentioned about parent/child relationships @alculquicondor  and make the broker pod a separate job that is launched first, and then have the other cluster workers somehow children? (is there an example of this somewhere I haven't done it before!) But that might be more effort / refactor than it's worth if we eventually want to go back to one IndexedJob with more structure.

File: site/content/en/docs/tasks/run_flux.md (hunk 9 above)
Comment (alculquicondor) [reply site/content/en/docs/tasks/run_flux.md:33]: For the purpose of queuing, we care about how many resources the pods will consume in the cluster.
We will obtain those from the created Job object.

File: site/content/en/docs/tasks/run_flux.md (hunk 2 above)
Comment (cortespao) [inline site/content/en/docs/tasks/run_flux.md:16]:
```suggestion
Check [Run A Job](/docs/tasks/run_jobs) for details on the initial cluster setup and queue configuration.
```

File: site/content/en/docs/tasks/run_flux.md (hunk 6 above)
Comment (cortespao) [inline site/content/en/docs/tasks/run_flux.md:67]:
```suggestion
- The certificate generation step of the MiniCluster life cycle runs outside its main job and its execution will not be managed by Kueue.
```

Assuming these are referring to the attributes that belong to the certificate generation

File: site/content/en/docs/tasks/run_flux.md (hunk 7 above)
Comment (cortespao) [inline site/content/en/docs/tasks/run_flux.md:68]: its instead of it's

File: site/content/en/docs/tasks/run_flux.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  jobLabels:
+    kueue.x-k8s.io/queue-name: user-queue
+```
+
+## 3. Known Limitations
+
+- The certificate generation step of the MiniCluster life cycle runs outside it's main job and it's execution will not be managed by Kueue.
+- In some configuration variants the MiniCluster's job might not run to completion blocking the resources of the associated queue until it's deletion.
+
+Both of them are under investigation and tracked by [150](https://github.com/flux-framework/flux-operator/issues/150) and [151](https://github.com/flux-framework/flux-operator/issues/151)
```
Comment (cortespao) [inline site/content/en/docs/tasks/run_flux.md:70]:
```suggestion
Both of these known issues are under investigation and tracked by [150](https://github.com/flux-framework/flux-operator/issues/150) and [151](https://github.com/flux-framework/flux-operator/issues/151)
```

**Review by cortespao — COMMENTED** [review]
A global suggestion is to unify the minicluster terms. The page has:
minicluster, Minicluster, Mini Cluster

Thanks :)

**Comment by vsoch** [comment]
> A global suggestion is to unify the minicluster terms. The page has: minicluster, Minicluster, Mini Cluster

The term I try to use most frequently is "MiniCluster" and that corresponds with the CRD kind.

File: site/content/en/docs/tasks/run_flux.md (hunk 9 above)
Comment (tenzen-y) [inline site/content/en/docs/tasks/run_flux.md:33]:
```suggestion
The resource needs of the workload can be configured in the `spec.container[*].resources` sections of the minicluster configuration.
```

It would be good to do the same as other places.
example: https://kueue.sigs.k8s.io/docs/tasks/administer_cluster_quotas/#2-create-a-clusterqueue-referencing-the-flavors

File: site/content/en/docs/tasks/run_flux.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
+---
+title: "Run A Flux Minicluster"
```
[the remark this answers is not in the document]
Comment (vsoch) [reply site/content/en/docs/tasks/run_flux.md:2]: yes correct - `MiniCluster` !

File: site/content/en/docs/tasks/run_flux.md (hunk 9 above)
Comment (vsoch) [reply site/content/en/docs/tasks/run_flux.md:33]: That might be hard to keep up to date / synced with our docs - so another suggestion is to just link to our documentation https://flux-framework.org/flux-operator/getting_started/custom-resource-definition.html#id3

Comment (tenzen-y) [reply site/content/en/docs/tasks/run_flux.md:33]: My suggestion is to add `*` to the array :)

Comment (tenzen-y) [reply site/content/en/docs/tasks/run_flux.md:33]: This means I didn't suggest referring to other documentation :)

File: site/content/en/docs/tasks/run_flux.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (vsoch) [reply site/content/en/docs/tasks/run_flux.md:2]: I'm going through our docs now to make sure all references are consistent - that will make it easier in the future to derive the right term.

File: site/content/en/docs/tasks/run_flux.md (hunk 9 above)
Comment (vsoch) [reply site/content/en/docs/tasks/run_flux.md:33]: Or sorry, it was hard to get the full context when you just said "do the same as other places" with a link. I did see the suggestion. And I think it still might be good to refer to the docs on our site!

File: site/content/en/docs/tasks/run_flux.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (trasc) [reply site/content/en/docs/tasks/run_flux.md:2]: done

File: site/content/en/docs/tasks/run_flux.md (hunk 8 above)
Comment (trasc) [reply site/content/en/docs/tasks/run_flux.md:56]: switch to `sleep 10`, to simulate some work

File: site/content/en/docs/tasks/run_flux.md (hunk 7 above)
Comment (trasc) [reply site/content/en/docs/tasks/run_flux.md:68]: we could leave this in and remove it later on

Comment (trasc) [reply site/content/en/docs/tasks/run_flux.md:68]: done

File: site/content/en/docs/tasks/run_flux.md (hunk 10 above)
Comment (trasc) [reply site/content/en/docs/tasks/run_flux.md:70]: removed

File: site/content/en/docs/tasks/run_flux_minicluster.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  Run a Kueue scheduled Flux MiniCluster.
+---
+
+This page shows how to leverage Kueue's scheduling and resource management capabilities when running [Flux Operator's](https://flux-framework.org/flux-operator/) MiniClusters.
+
+This guide is for [batch users](/docs/tasks#batch-user) that have a basic understanding of Kueue. For more information, see [Kueue's overview](/docs/overview).
+
+## Before you begin
+
+Check [Run A Job](/docs/tasks/run_jobs) for details on the initial cluster setup and [queue configuration](/docs/tasks/administer_cluster_quotas).
```
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux_minicluster.md:15]:
```suggestion
Check [administer cluster quotas](/docs/tasks/administer_cluster_quotas) for details on the initial cluster setup.
```

Comment (alculquicondor) [reply site/content/en/docs/tasks/run_flux_minicluster.md:15]: Also add a link to the guide to install flux-operator

File: site/content/en/docs/tasks/run_flux_minicluster.md [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+## 1. MiniCluster definition
```
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux_minicluster.md:17]:
```suggestion
## MiniCluster definition
```

File: site/content/en/docs/tasks/run_flux_minicluster.md [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+spec:
+  containers:
+    - image: <image>
+      resources:
+        requests:
+          cpu: 4
+          memory: "200Mi"
+```
+
+## 2. Sample MiniCluster
```
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux_minicluster.md:45]:
```suggestion
## Sample MiniCluster
```

**Review by alculquicondor — COMMENTED** [review]
@vsoch are you ok with us publishing this doc with the limitation about termination?

**Comment by alculquicondor** [comment]
yes, that's what I meant. Then this can be closed? https://github.com/flux-framework/flux-operator/issues/151

File: site/content/en/docs/tasks/run_flux_minicluster.md (hunk 12 above)
Comment (trasc) [reply site/content/en/docs/tasks/run_flux_minicluster.md:15]: done

File: site/content/en/docs/tasks/run_flux_minicluster.md (hunk 13 above)
Comment (trasc) [reply site/content/en/docs/tasks/run_flux_minicluster.md:17]: done

File: site/content/en/docs/tasks/run_flux_minicluster.md (hunk 14 above)
Comment (trasc) [reply site/content/en/docs/tasks/run_flux_minicluster.md:45]: done

File: site/content/en/docs/tasks/run_flux.md (hunk 7 above)
Comment (trasc) [reply site/content/en/docs/tasks/run_flux.md]: removed

File: site/content/en/docs/tasks/run_flux_minicluster.md [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+This page shows how to leverage Kueue's scheduling and resource management capabilities when running [Flux Operator's](https://flux-framework.org/flux-operator/) MiniClusters.
+
+This guide is for [batch users](/docs/tasks#batch-user) that have a basic understanding of Kueue. For more information, see [Kueue's overview](/docs/overview).
+
+## Before you begin
+
+Check [administer cluster quotas](/docs/tasks/administer_cluster_quotas) for details on the initial cluster setup.
+
+Check its [getting started](https://flux-framework.org/flux-operator/getting_started/user-guide.html#install) for details on Flux Operator installation
```
Comment (alculquicondor) [inline site/content/en/docs/tasks/run_flux_minicluster.md:17]:
```suggestion
Check [the Flux Operator installation guide](https://flux-framework.org/flux-operator/getting_started/user-guide.html#install).
```

**Review by alculquicondor — COMMENTED** [review]
/approve

File: site/content/en/docs/tasks/run_flux_minicluster.md (hunk 15 above)
Comment (trasc) [reply site/content/en/docs/tasks/run_flux_minicluster.md]: done

**Comment by vsoch** [comment]
@trasc would you like me to run through a tutorial / test soon? It would be good for me to try Kueue anyway!

**Comment by trasc** [comment]
> @trasc would you like me to run through a tutorial / test soon? It would be good for me to try Kueue anyway!

Sure

Note that we now have [helm support](https://github.com/kubernetes-sigs/kueue/tree/main/charts/kueue) which might help wit the  installation. 

Since the sample queue has a 9 cpu quota, creating multiple `Sample MiniCluster` you should see at most tow running at the same time.

**Comment by alculquicondor** [comment]
/approve
/lgtm

Thanks!
